### PR TITLE
MAINT: rename IOError -> OSError

### PR DESCRIPTION
**Pull request by mwtoews**
Since Python 3.3,  IO and OS exceptions were re-worked as per [PEP 3151](https://www.python.org/dev/peps/pep-3151/). The new main base class is [OSError](https://docs.python.org/3/library/exceptions.html#OSError), with several other exceptions merged, including IOError (i.e. `assert IOError is OSError`).

This PR changes many of the old alias to OSError, with a few exceptions:

- Versioneer EnvironmentError were not modified (this change has been merged upstream)
- In `pandas/_libs/parsers.pyx`, this is not an OSError, as there is no issue reading the file. The exception is reclassified as a [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError).

Are any of these changes worthy of a whatsnew entry?

### CODE REVIEW

**Review by twoertwein — APPROVED** [review]
Thank you! The changes look good to me.

**Comment by jreback** [comment]
@mwtoews can you add a whatsnew note in 1.4., I/O Bug fix section is fine, just noting that we changed this. (and ref this issue). pls merge master and ping on green.

**Comment by mwtoews** [comment]
@jreback added a whatsnew entry to describe the user-visible change, let me know if it suites or needs rephrasing. I don't understand the Azure "The operation was canceled" CI failure.

**Review by lithomas1 — CHANGES_REQUESTED** [review]
Can you resolve conflicts? Don't worry about the timeouts , they're flaky.

File: doc/source/whatsnew/v1.4.0.rst [hunk 1]
```diff
@@ -385,6 +385,7 @@ I/O
 - Column headers are dropped when constructing a :class:`DataFrame` from a sqlalchemy's ``Row`` object (:issue:`40682`)
 - Bug in unpickling a :class:`Index` with object dtype incorrectly inferring numeric dtypes (:issue:`43188`)
 - Bug in :func:`read_csv` where reading multi-header input with unequal lengths incorrectly raising uncontrolled ``IndexError`` (:issue:`43102`)
+- Bug in ``TextReader`` (used for text-based read methods), changed exception class when expecting a file path name or file-like object from ``IOError`` (alias of ``OSError``) to ``TypeError`` (:issue:`43366`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.4.0.rst:388]: just say `pd.read_csv` as that will render (TextReader will not like you have it here)

File: pandas/_libs/parsers.pyx [hunk 2]
```diff
@@ -607,8 +607,8 @@ cdef class TextReader:
             void *ptr
 
         if not hasattr(source, "read"):
-            raise IOError(f'Expected file path name or file-like object, '
-                          f'got {type(source)} type')
+            raise TypeError('Expected file path name or file-like object, '
```
Comment (jreback) [inline pandas/_libs/parsers.pyx:610]: can you construct a test that actually hits this? (e.g. you didn't have to change any tests here)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by mwtoews** [comment]
The scope of this PR has been reduce to a pure maintenance change, which shouldn't need any whatsnew entry, since any IOError class thrown was viewed as OSError to the user, so no changes should be visible.

The more involved bug fix that changes the exception class, and works across different parsers is at #43443

File: pandas/_libs/parsers.pyx (hunk 2 above)
Comment (mwtoews) [reply pandas/_libs/parsers.pyx:610]: Added `pandas/tests/io/parser/common/test_common_basic.py::test_read_filepath_or_buffer` which checks four parsers.

File: doc/source/whatsnew/v1.4.0.rst (hunk 1 above)
Comment (mwtoews) [reply doc/source/whatsnew/v1.4.0.rst]: done

File: pandas/io/common.py [hunk 3]
```diff
@@ -738,6 +738,10 @@ def get_handle(
             isinstance(ioargs.filepath_or_buffer, str) or ioargs.should_close
         )
 
+    if not hasattr(handle, "read"):
```
Comment (twoertwein) [inline pandas/io/common.py:741]: You probably need to change the condition to `"r" in ioargs.mode and not hasattr(handle, "read")`.

If you have the check here, I don't think the code in pandas/_libs/parsers.pyx will ever be reached.

Comment (mwtoews) [reply pandas/io/common.py]: Thanks changes done. It seems that the other check can indeed be removed, as all parsers use `get_handle`.

**Comment by alimcmaster1** [comment]
@mwtoews - can you take a look at the test failures?

**Comment by mwtoews** [comment]
The third commit in this PR fixes the excel readers, which allow bytes for the first parameter (e.g. [pandas.read_excel](https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html)). Since `get_handle` expects FilePathOrBuffer, these bytes need to be passed as BytesIO buffers.

I'd appreciate a review from anyone familiar with `pandas/io/excel/_base.py`. As these are internal changes to the excel readers, I'm not expecting any further whatsnew notes on this aspect, but please correct me if it should.

File: pandas/io/excel/_base.py [hunk 4]
```diff
@@ -1170,6 +1172,10 @@ def __init__(
         if engine is not None and engine not in self._engines:
             raise ValueError(f"Unknown engine: {engine}")
 
+        # First argument can also be bytes, so create a buffer
```
Comment (jreback) [inline pandas/io/excel/_base.py:1175]: is there a test that hits this?

**Review by jreback — CHANGES_REQUESTED** [review]
cc @rhshadrach

File: pandas/io/excel/_base.py (hunk 4 above)
Comment (mwtoews) [reply pandas/io/excel/_base.py:1175]: yes there are several, both failing and passing. E.g. `pandas/tests/io/excel/test_readers.py::TestReaders::test_corrupt_bytes_raises`

**Comment by mwtoews** [comment]
It seems that support for reading bytes for the excel readers was added with #30519

Also, feel free to re-title and re-classify this PR, as the scope has shifted since initially submitted.

**Comment by jreback** [comment]
> It seems that support for reading bytes for the excel readers was added with #30519
> 
> Also, feel free to re-title and re-classify this PR, as the scope has shifted since initially submitted.

no its fine, just the question on where the bytes are wrapped is still likely hit

**Review by rhshadrach — COMMENTED** [review]
excel changes lgtm - verified tests would fail without the changes in `excel/_base`

**Comment by jreback** [comment]
pre commit issue: https://github.com/pandas-dev/pandas/pull/43366/checks?check_run_id=3555699486

**Review by jreback — APPROVED** [review]

**Comment by lithomas1** [comment]
@github-actions pre-commit

**Review by lithomas1 — APPROVED** [review]
LGTM. and thanks for the PR (I took care of the pre-commit issue for you).

P.S. You don't have to force-push/squash your commits if you don't want to. We squash on merge so it's fine.

**Comment by lithomas1** [comment]
thanks @mwtoews

**Pull request merged**